File: Chapter 1/Section 1.6/numtri/OLD/numtri_OLD.py
"""
ID: ten.to.1
TASK: numtri
LANG: PYTHON3
"""

class TriNode:
	def __init__(self, val):
		self.value = val
		self.right = None
		self.left  = None

f_in  = open("numtri.in", "r");
f_out = open("numtri.out", "w")
R     = int(f_in.readline())

nodes = []

for i in range(0, R):
	nodes.append(list(map(TriNode, map(int, f_in.readline().split()))))

for i in range(0, R - 1):
	for j in range(0, i + 1):
		nodes[i][j].left  = nodes[i + 1][j]
		nodes[i][j].right = nodes[i + 1][j + 1]

#Print Nodes
#for n in nodes:
#	for j in n:
#		print(j.value, end=" ")
#	print()

max = 0

def solve(node, sum):
	sum += node.value;
	
	if(node.left == None):
		if(sum > max):
			max = sum
		return
	
	solve(node.left, sum)
	solve(node.right, sum)

solve(nodes[0][0], 0)

f_out.write(str(max) + "\n")

File: Chapter 1/Section 1.6/numtri/OLD/numtri_OLD.cpp
/*
ID: ten.to.1
TASK: numtri
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <algorithm>
#include <cmath>
//#include <vector>

struct TriNode
{
	int value;
	TriNode* left;
	TriNode* right;
	
	TriNode()
	{
		
	}
	
	TriNode(int v)
	{
		value = v;
		left  = NULL;
		right = NULL;
	}
};

int sum_fact(int n)
{
	return (n == 0) ? 0 : sum_fact(n - 1) + n;
}

void solve(TriNode* node, int& max, int sum)
{
	sum += node->value;
	
	if(node->left == NULL)
	{
		std::cout << sum << '\n';
		if(max < sum)
		{
			max = sum;
		}
		return;
	}
	
	solve(node->left, max, sum);
	solve(node->right, max, sum);
}

int main()
{
	std::ifstream filein("numtri.in");
	std::ofstream fileout("numtri.out");
	
	int R; filein >> R;
	int num_of_nodes = sum_fact(R);
	TriNode nodes[R][num_of_nodes];
	//std::vector<std::vector<TriNode>> nodes;
	
	for(int i = 0; i < R; i++)
	{
		//std::vector<TriNode> temp;
		for(int j = 0; j < i + 1; j++)
		{
			int v; filein >> v;
			TriNode t(v);
			nodes[i][j] = t;
			
			/*
			int v; filein >>v ;
			TriNode t(v);
			temp.push_back(t);
			*/
		}
		//nodes.push_back(temp);
	}
	
	for(int i = 0; i < (R - 1); i++)
	{
		for(int j = 0; j < i + 1; j++)
		{
			nodes[i][j].left  = &nodes[i + 1][j];
			nodes[i][j].right = &nodes[i + 1][j + 1];
		}
	}
	int max = -1;
	solve(&nodes[0][0], max, 0);
	fileout << max << '\n';
}

